(1322 part 1) add visualiser option to generate command & remove separate visualise command

afroggattsl opened the pull request:
### Description
First two tasks of #1322 which is to add ability to turn on visualisation when using generate command and to remove the separate visualise command

### Changes
* can now set a visualiser-level (OFF by default) that means get  .dot files generated with graphs of the decision trees
* currently just produces initial tree and tree after up-front pruning
* there is a visualiser-output-folder option too which defaults to current directory
* added tests for all new classes and updated tests of any classes changed
* removed VisualiseExecute, VisualiseCommandLine classes and references from App class

### Additional notes
* leaving old way of doing visualisation for now
* graph viz tool only handles one graph per file hence need ot generate multiple files.  Only 2 for thsi PR but will be more when incorporate tree printing elsewhere in code

### Issue
Related to #1322

pdaulbyscottlogic commented on orchestrator/src/main/java/com/scottlogic/deg/orchestrator/generate/GenerateCommandLine.java:
```diff
@@ -105,51 +109,59 @@ private void checkForAlphaGenerationDataTypes() {
         description = "Defines the maximum number of rows that should be generated")
     private long maxRows = DEFAULT_MAX_ROWS;
 
+    @SuppressWarnings("FieldCanBeLocal")
```
i don't get any warnings for these, does it need to be supressed? what's the warning about?
can it be supressed at class level, rather than on each field?

afroggattsl replied:
warns me field can be converted to a local variable.  maybe you have added some suppressions to your local intellij project files?

pdaulbyscottlogic commented on generator/src/main/java/com/scottlogic/deg/generator/config/detail/VisualiserLevel.java:
```diff
@@ -0,0 +1,26 @@
+        }
+    };
+
+    public abstract boolean sameOrHigherThan(VisualiserLevel level);
```
this method seems more confusing than not having the method. it's only used in one place

afroggattsl replied:
just doing one implementation and no overriding as its still small.  Will rename it so more obvious we want to know if level is same level of verbosity or more verbose than level passed in.  So its just:
    public boolean sameOrMoreVerboseThan(VisualiserLevel level) {
        return level == OFF || level == STANDARD && this != OFF || level == this;
    }

pdaulbyscottlogic commented on generator/src/main/java/com/scottlogic/deg/generator/generation/DecisionTreeDataGenerator.java:
```diff
@@ -81,4 +92,12 @@ public DecisionTreeDataGenerator(
+    private void visualiseTree(DecisionTree decisionTree, String title) {
+        try (Visualiser visualiser = visualiserFactory.create(VisualiserLevel.STANDARD, title)) {
+            visualiser.printTree(title, decisionTree);
+        } catch (Exception e) {
```
is it possible to catch these in the visualisers?

afroggattsl replied:
i will catch and rethrow as runtime exceptions in visualiser.printTree so callers don't need to do and trys.  likewise i have to remove ioexception from Autocloseable::close() method it needs to implement (that implemented so can be used in a  try-resource block) - and there will need to be a catch and re-throw as runtime as underlying string writer's close method will throw IOexception.  but again will do thsi lower down so callers like thsi class don't need to handle it

pdaulbyscottlogic commented on generator/src/main/java/com/scottlogic/deg/generator/config/detail/VisualiserLevel.java:
```diff
@@ -0,0 +1,26 @@
+package com.scottlogic.deg.generator.config.detail;
+
+public enum VisualiserLevel  {
```
this never really seems to be used, is it removable?

afroggattsl replied:
this class is used.  if user does not specify --visualiser-level it defaults to OFF, otherwise if they specify standard it will log out the initial, pruned, optimised and partitioned trees.  if specify DETAILED it will also print out trees throughout each prune done during the walking.  This PR does not do all of this yet.  This will be in follow up PRs.  This PR will include hwta done so far, the fixes you suggested and updates to the DOCs (which not done yet)

pdaulbyscottlogic commented on generator/src/main/java/com/scottlogic/deg/generator/guice/VisualiserFactoryProvider.java:
```diff
@@ -0,0 +1,23 @@
+import com.scottlogic.deg.generator.generation.visualiser.VisualiserFactory;
+import com.scottlogic.deg.generator.generation.visualiser.VisualiserWriterFactory;
+
+public class VisualiserFactoryProvider implements Provider<VisualiserFactory> {
```
visualiser factory seems like it could be created without the need for a provider

afroggattsl replied:
yes it can.  will get rid of

afroggattsl commented:
> Seems to be in a weird spot, as it's replacing something else, but not quite finishing everything visualisation related.
> 
> I've noted some things that may be changable, but don't know if they'd get changed again in the future
> 
> That said, i think the docs needs updating, especially if visualisation level stays in its current state.

indeed. i possibly should just have done all in one PR but I have split it up.  I will add docs to this PR and then will do another PR to finish it off (ie add it into rest of the places and update the docs again).  just trying to reduce number of things changing in a PR.  but multiple PRs brings its own issues.

afroggattsl commented:
commit 4 (with "fixes from first round of code review" in the commit message body) addresses comments made by paul in first round of code review.  will do another commit today/tomorrow which updates the docs.  Again, there will be another PR (under same owning issue #1322) to add visualisation to later parts of pipeline and the walker

afroggattsl commented:
> Would it be possible for you to add some changes to documentation?

yep, its on my list, just not had time yet
